_posts/2020-02-11-spring-boot-openapi.md

```diff
@@ -0,0 +1,394 @@
+            write:users: modify users
+            read:users: read users
+    api_key:
+      type: apiKey
+      name: api_key
+      in: header
+```
+
+* First the [`schemas`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#schema-object) field, allows the definition of our DTOs.
+* And last the [`securitySchemes`](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#security-scheme-object), and it defines a security scheme that can be used by the operations.
+
+## Generating Code From an API Specification
+
+Having defined an API, we'll now create code from the YAML document above. 
+
+We'll take a look at two different approaches to generating the code:
+
+1. Using [Swagger Editor](http://editor.swagger.io/) to generate code manually, and
+2. Using the [OpenAPI Maven plugin](https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator-maven-plugin) to generate code from a Maven build.
+
+### Generating Code from Swagger Editor
+
+Although that's an approach which I wouldn't take, I must introduce it. 
+
+So, let's go over to [Swagger Editor](http://editor.swagger.io/) and paste our `<specification>.yml` into it. Then, we select *Generate Server* from the menu and pick what kind of a server we'd would like to generate (I went with Spring).
+
+As I said before, I wouldn't use this approach. But why?
+
+First, the code that was generated for me is using Java 7 and Spring Boot 1.5.22.RELEASE, both of which are quite old now. 
+
+Second, if we make a change to the specification (and changes happen all the time), we'd have to copy-and-paste the files that were changed. 
+
+ ### Generating Code with the OpenAPI Maven plugin
+ 
+ A better alternative is to generate the code from within a Maven build with the [OpenAPI Maven plugin](https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator-maven-plugin). 
+ 
+  Let's take a look at the folder structure. I chose to use a multi-module maven project, where we have two projects:
+ 
+ 1. `app`, an application that implements the API from our specification.
+ 2. `specification`, whose only job is to provide the API Specification for our app.  
+ 
+ The folder structure looks like this:
+ 
+```text
+spring-boot-openapi
+├── app
+│   └── pom.xml
+│   └── src
+│       └── main
+│           └── java
+│               └── io.reflectoring
+│                   └── OpenAPIConsumerApp.java
+├── specification
+│   └── pom.xml
+│   └── src
+│       └── resources
+│           └── openapi.yml
+├── pom.xml
+```
+
+For the shake of simplicity, we omit the test folders.
+
+Our `app` is a simple Spring Boot project that we can automatically generate on [start.spring.io](https://start.spring.io), so let's focus on the `pom.xml` from the `specification` module, where we configure 
+the [OpenAPI Maven plugin](https://github.com/OpenAPITools/openapi-generator/tree/master/modules/openapi-generator-maven-plugin):
```

Comment:
We don't need to link it always...the first time is enough. Otherwise the text is riddled with links. Can you please remove all occurrences of links exception the first one (not only for the Plugin, but please check for other links as well).